### ISSUE TITLE
Unable to deserialize nested sealed classes

### ISSUE THREAD

**Issue by dualscyther** [Kotlin/kotlinx.serialization#126]
```kt
@Serializable
data class FooHolder(val someMetadata: Int, val payload: Foo)
sealed class Foo
@Serializable
data class Bar(val bar: Int) : Foo()
@Serializable
data class Baz(val baz: Int) : Foo()
```
...

```kt
val msgString = JSON.stringify(FooHolder.serializer(), FooHolder(0, Bar(1)))
val msgString2 = JSON.stringify(FooHolder.serializer(), FooHolder(0, Baz(1)))
println(msgString)
println(msgString2)
val msg = JSON.parse(FooHolder.serializer(), msgString)
val msg2 = JSON.parse(FooHolder.serializer(), msgString2)
println(msg)
println(msg2)
```

The above code fails at the `JSON.parse` with

> kotlinx.serialization.SerializationException: Any type is not supported

even though printing `msgString` and `msgString2` succeeds, with output:

> `{"someMetadata":0,"payload":"Bar(bar=1)"}`
> `{"someMetadata":0,"payload":"Baz(baz=1)"}`

I assume this is something to do with the deserializer getting stuck on instantiating a `Foo`, but I'm not really sure how to fix it. I do know that if I try to annotate the `Foo` class itself with `@Serializable` then I get:

> `java.lang.IndexOutOfBoundsException: Index: 0, Size: 0`

**Comment by pdvrieze** [comment]
I suspect very strongly that this is caused by the incorrect generation of enum serialization code: See #125

**Comment by dualscyther** [comment]
@sandwwraith nice, giving special treatment to `val payload` fixes it:

```kt
@Serializable
data class FooHolder(
    val someMetadata: Int,
    @Serializable(with = PolymorphicSerializer::class) val payload: Foo
)
```

But if `Bar` or `Baz` are generic then there is no way to specify the generic serializer. This is the case even if `FooHolder` and `Foo` are also generic - there is no way to pass down the generic serializer through a tree of wrapper classes (since companion object customization of the default serializer isn't yet available for generic classes).

**Comment by ideaBlender** [comment]
Just wondering on the current state of serializaing sealed classes.

would serializing something like 
```
sealed class ServiceResult<T> {
    class Failure<T>(val errMsg: String) : ServiceResult<T>()
    class Success<T>(val payload: T) : ServiceResult<T>()
}
```

be a problem?

**Comment by Whathecode** [comment]
I believe this to be fixed with the 0.14.0 release which now fully supports sealed classes? I particularly like how extending classes no longer need to be registered as polymorphic in a serial module.

Great job, guys!

**Comment by sandwwraith** [comment]
Thanks!

I think so. If it is still a problem, feel free to reopen.
